_posts/2020-03-16-dask-usage-at-blue-yonder.markdown

```diff
@@ -0,0 +1,161 @@
+the case that we have the data available in a suitable partitioning, for instance, if the data results
+from a computation that used a different partitioning. In this case, we have to repartition the data.
+This works by reading the dataset as a [Dask dataframe](https://docs.dask.org/en/latest/dataframe.html) 
+repartitioning the dataframe using network shuffle, and writing it out again to a dataset.
+
+## Ad hoc data analysis
+
+Finally, our data scientists also use Dask for ad hoc analysis of data that is too large to
+fit into an individual compute node's memory.
+
+## Dask cluster setup at Blue Yonder
+
+We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
+of our domain and partly historically grown. We are running a large number of individual clusters
+that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
+setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number
+of upstream pull requests.
+
+### Cluster separation
+
+We are running separate Dask clusters for separate customers in order to isolate the customers from each other.
+For the same reason, we also run separate clusters for production, staging, and development environments. 
+
+But it does not stop there. Our compute pipeline involves different steps, for instance, downloading the input data
+ into a datasets, calculating predictions from the input data, calculating order proposals from the predictions, etc.
+The different steps involve software that is written and maintained by different teams.
+For performance reasons, we install the Python packages holding the code needed
+ for the computations on each worker. Since we do not want to synchronize the Python packages required for each pipeline
+ step (e.g., allow different Pandas versions for calculating order proposals and predictions in order not to introduce
+ a coupling of the software release cycles), we have to run a separate Dask cluster for each of pipeline steps. 
+This means we operate more than
+ten Dask clusters per customer and environment, with most of the time, only one of the clusters being active
+and computing something. While this leads to overhead in terms of administration and hardware resouces,
+it also gives us a lot of flexibility. For instance, we can update the software on the cluster of one part of the compute pipeline
+while another part of the pipeline is computing something on a different cluster.
+
+### Some numbers
+
+The number and size of the workers varies from cluster to cluster depending on the degree of parallelism
+of the computation being performed, its resource requirements, and the available timeframe for the computation.
+At the time of writing this, we are running more than 500 distinct clusters.
+They have between one and 225 workers, with worker size varying between 1GiB and 64GiB of memory.
+Our total memory use goes up to as much as 15TiB.
```

Comment:
how many CPUs? 

Can we visualise this somehow (if there is no sane visualization, that's also fine, of course)?